prototypes/pipemarker.lua

```diff
@@ -0,0 +1,149 @@
+local merge = _G.util.merge
+local base_entity = {
+    type = 'corpse',
+    name = 'fillerstuff',
+    flags = {'placeable-neutral', 'not-on-map'},
+    subgroup = 'remnants',
+    order = 'd[remnants]-c[wall]',
+    icon = '__core__/graphics/empty.png',
+    icon_size = 1,
+    time_before_removed = 2000000000,
+    collision_box = {{0, 0}, {0, 0}},
+    selection_box = {{-0.5, -0.5}, {0.5, 0.5}},
+    selectable_in_game = false,
+    final_render_layer = 'selection-box',
+    animation = {
+        width = 64,
+        height = 64,
+        frame_count = 1,
+        direction_count = 1,
+        shift = {-0.5, -0.5},
+        filename = '__pipelayer__/graphics/entity/markers/pipe-marker-dot.png'
+    }
+}
+
+local directional_table = {
+    ['0'] = '',
+    [1] = '-n',
+    [4] = '-e',
+    [5] = '-ne',
+    [16] = '-s',
+    [17] = '-ns',
+    [20] = '-se',
+    [21] = '-nse',
+    [64] = '-w',
+    [65] = '-nw',
+    [68] = '-ew',
+    [69] = '-new',
+    [80] = '-sw',
+    [81] = '-nsw',
+    [84] = '-sew',
+    [85] = '-nsew'
+}
+local new_dots = {}
+for direction_index, directions in pairs(directional_table) do
+    local current_entity = util.table.deepcopy(base_entity)
+    --current_entity.type = 'corpse'
+    current_entity.name = 'pipelayer-pipe-dot' .. directions
+    if direction_index == '0' then
+        current_entity.final_render_layer = 'light-effect'
+        --? The single pipe dot is twice as large so must be scaled
+        current_entity.animation.scale = 0.5
+    else
+      current_entity.animation.width = 32
+      current_entity.animation.height = 32
+    end
+    current_entity.animation.filename = '__pipelayer__/graphics/entity/markers/pipe-marker-dot' .. directions .. '.png'
+    new_dots[#new_dots + 1] = current_entity
+end
+
+for _, stuff in pairs(new_dots) do
+  data:extend {
+      merge {
```

Comment:
You're already doing a deep copy on line 45, so I'm not sure why the merge. I think this could just be `data:extend(new_dots)`, or even just call `data:extend` from inside the for loop.

Reply:
Had a weird kink in mine because it wouldn't work right other wise so I just left it alone. Didn't dig into it. If you end up sorting it I would gladly accept the knowledge.